internal/server/data/identity.go

```diff
@@ -161,31 +163,63 @@ func DeleteIdentities(tx GormTxn, selectors ...SelectorFunc) error {
 		return err
 	}
 
-	ids := make([]uid.ID, 0)
-	for _, i := range toDelete {
-		ids = append(ids, i.ID)
+	ids, err := deleteReferencesToIdentities(tx, providerID, toDelete)
+	if err != nil {
+		return fmt.Errorf("remove identities: %w", err)
+	}
 
-		err := DeleteGrants(tx, DeleteGrantsOptions{BySubject: i.PolyID()})
-		if err != nil {
-			return err
+	if len(ids) > 0 {
+		return deleteAll[models.Identity](tx, ByIDs(ids))
+	}
+
+	return nil
+}
+
+func deleteReferencesToIdentities(tx GormTxn, providerID uid.ID, toDelete []models.Identity) (unreferencedIdentityIDs []uid.ID, err error) {
+	for _, i := range toDelete {
+		if err := DeleteAccessKeys(tx, DeleteAccessKeysOptions{ByIssuedForID: i.ID, ByProviderID: providerID}); err != nil {
+			return []uid.ID{}, fmt.Errorf("delete identity access keys: %w", err)
+		}
+		if providerID == InfraProvider(tx).ID {
+			// if an identity does not have credentials in the Infra provider this won't be found, but we can proceed
+			credential, err := GetCredential(tx, ByIdentityID(i.ID))
+			if err != nil && !errors.Is(err, internal.ErrNotFound) {
+				return []uid.ID{}, fmt.Errorf("get delete identity creds: %w", err)
+			}
+			if credential != nil {
+				err := DeleteCredential(tx, credential.ID)
+				if err != nil {
+					return []uid.ID{}, fmt.Errorf("delete identity creds: %w", err)
+				}
+			}
+		}
+		if err := DeleteProviderUsers(tx, DeleteProviderUsersOptions{ByIdentityID: i.ID, ByProviderID: providerID}); err != nil {
+			return []uid.ID{}, fmt.Errorf("remove provider user: %w", err)
 		}
 
-		groups, err := ListGroups(tx, nil, ByGroupMember(i.ID))
+		// if this identity only exists from an infra identity provider, remove all their groups and grants, then delete the user
+		user, err := GetIdentity(tx, Preload("Providers"), ByID(i.ID))
 		if err != nil {
-			return err
+			return []uid.ID{}, fmt.Errorf("check user providers: %w", err)
 		}
 
-		for _, group := range groups {
-			err = RemoveUsersFromGroup(tx, group.ID, []uid.ID{i.ID})
+		if len(user.Providers) == 0 {
```

Comment:
I don't think I follow this comment. This happens when they are removed from the last provider (regardless of which provide it was), doesn't it? Or because we don't call this from `UpdateIdentityInfoFromProvider` then presumably the last provider was infra?

Reply:
This was a bad comment, fixed